src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java

```diff
@@ -129,33 +267,437 @@ public ResultSet getSchemas() throws SQLException {
+                new BsonElement(IS_AUTOINCREMENT, new BsonString("")),
+                new BsonElement(IS_GENERATEDCOLUMN, new BsonInt32(0)));
+    }
+
+    // Helper for creating BSON documents for the getColumnPrivileges methods. Intended
+    // for use with the getColumnsFromDB helper method which is shared between getColumns
+    // and getColumnPrivileges.
+    private BsonDocument toGetColumnPrivilegesDoc(GetColumnsDocInfo i) {
+        return createBottomBson(
+                new BsonElement(TABLE_CAT, new BsonString(i.dbName)),
+                new BsonElement(TABLE_SCHEM, new BsonString("")),
+                new BsonElement(TABLE_NAME, new BsonString(i.collName)),
+                new BsonElement(COLUMN_NAME, new BsonString(i.columnName)),
+                new BsonElement(GRANTOR, new BsonNull()),
+                new BsonElement(GRANTEE, new BsonString("")),
+                new BsonElement(PRIVILEGE, new BsonString("SELECT")),
+                new BsonElement(IS_GRANTABLE, new BsonNull()));
+    }
+
+    // Helper for getting column data for all columns from all tables from a specific
+    // database. Used by getColumns and getColumnPrivileges. The caller specifies how
+    // to serialize the column info into BSON documents for the result set.
+    private Stream<BsonDocument> getColumnsFromDB(
+            String dbName,
+            Pattern tableNamePatternRE,
+            Pattern columnNamePatternRE,
+            Function<GetColumnsDocInfo, BsonDocument> f) {
+        MongoDatabase db = this.conn.getDatabase(dbName).withCodecRegistry(MongoDriver.registry);
+
+        return db.listCollectionNames()
+                .into(new ArrayList<>())
+                .stream()
+
+                // filter only for collections matching the pattern
+                .filter(collName -> tableNamePatternRE.matcher(collName).matches())
+
+                // map the collection names into triples of (dbName, collName, collSchema)
+                .map(
+                        collName ->
+                                new Pair<>(
+                                        new Pair<>(dbName, collName),
+                                        db.runCommand(
+                                                new BsonDocument(
+                                                        "sqlGetSchema", new BsonString(collName)),
+                                                MongoJsonSchemaResult.class)))
+
+                // filter only for collections that have schemas
+                .filter(p -> p.right().ok == 1 && p.right().schema.jsonSchema.isObject())
+
+                // flatMap the column data into a single stream of BSON docs
+                .flatMap(
+                        p -> {
+                            Pair<String, String> ns = p.left();
+                            MongoJsonSchemaResult res = p.right();
+                            AtomicInteger idx = new AtomicInteger();
```

Comment:
Used for tracking ordinal position (note that we base position on lexicographic sort order). Should I say that in a code comment?

Reply:
Not sure why you need an AtomicInteger here? This is not multithreaded, is it (I don't think so otherwise you would potentially loose the sorting)?

Reply:
it is not, I could just use a normal int and `++` I think

Reply:
ah this is because of the compiler error "Variable used in lambda expression should be final or effectively final." My IDE suggested using `AtomicInteger` or using a 1-element array, since making this a `final int` would not be helpful. `AtomicInteger`, while overkill for a single-threaded use case, is still useful here simply because it is an int wrapper that has the `++` functionality we need here.

Reply:
Ah yes, this is because Java's safety algorithm is overly conservative.